File: go/scorers_test.go
package evalforge

import (
	"math"
	"testing"
)

func approxEqual(t *testing.T, got, want float64) {
	t.Helper()
	if math.Abs(got-want) > 1e-9 {
		t.Fatalf("got %v, want %v", got, want)
	}
}

func TestExactMatchIdentical(t *testing.T) {
	if got := ExactMatch("abc", "abc"); got != 1.0 {
		t.Fatalf("ExactMatch identical = %v, want 1.0", got)
	}
}

func TestExactMatchTrimsSurroundingWhitespace(t *testing.T) {
	if got := ExactMatch("  abc\n", "abc"); got != 1.0 {
		t.Fatalf("ExactMatch with surrounding whitespace = %v, want 1.0", got)
	}
}

func TestExactMatchDifferent(t *testing.T) {
	if got := ExactMatch("abc", "abd"); got != 0.0 {
		t.Fatalf("ExactMatch different = %v, want 0.0", got)
	}
}

func TestExactMatchIsCaseSensitive(t *testing.T) {
	if got := ExactMatch("ABC", "abc"); got != 0.0 {
		t.Fatalf("ExactMatch is case sensitive, got %v, want 0.0", got)
	}
}

func TestContainsSubstring(t *testing.T) {
	if got := Contains("the answer is 30 days", "30 days"); got != 1.0 {
		t.Fatalf("Contains substring = %v, want 1.0", got)
	}
}

func TestContainsIsCaseInsensitive(t *testing.T) {
	if got := Contains("The Answer Is 30 DAYS", "30 days"); got != 1.0 {
		t.Fatalf("Contains should be case-insensitive, got %v, want 1.0", got)
	}
}

func TestContainsTrimsExpected(t *testing.T) {
	if got := Contains("abc def ghi", "  def  "); got != 1.0 {
		t.Fatalf("Contains should trim expected, got %v, want 1.0", got)
	}
}

func TestContainsAbsent(t *testing.T) {
	if got := Contains("hello world", "goodbye"); got != 0.0 {
		t.Fatalf("Contains absent = %v, want 0.0", got)
	}
}

func TestRegexMatches(t *testing.T) {
	if got := RegexMatch("No discount", "(?i)no"); got != 1.0 {
		t.Fatalf("RegexMatch (?i)no = %v, want 1.0", got)
	}
}

func TestRegexInlineCaseInsensitiveAlternation(t *testing.T) {
	if got := RegexMatch("No, we do not offer a 90% loyalty discount.", "(?i)(no|not|don't|do not)"); got != 1.0 {
		t.Fatalf("RegexMatch alternation = %v, want 1.0", got)
	}
}

func TestRegexNoMatch(t *testing.T) {
	if got := RegexMatch("yes absolutely", "(?i)no"); got != 0.0 {
		t.Fatalf("RegexMatch no match = %v, want 0.0", got)
	}
}

func TestRegexSearchesAnywhere(t *testing.T) {
	if got := RegexMatch("the total is 30 days", `\d+ days`); got != 1.0 {
		t.Fatalf("RegexMatch should search anywhere, got %v, want 1.0", got)
	}
}

func TestSemanticIdenticalIsOne(t *testing.T) {
	approxEqual(t, SemanticSimilarity("abc def", "abc def"), 1.0)
}

func TestSemanticEmptyOutputIsZero(t *testing.T) {
	approxEqual(t, SemanticSimilarity("", "something"), 0.0)
}

func TestSemanticEmptyExpectedIsZero(t *testing.T) {
	approxEqual(t, SemanticSimilarity("something", ""), 0.0)
}

func TestSemanticBothEmptyIsZero(t *testing.T) {
	approxEqual(t, SemanticSimilarity("", ""), 0.0)
}

// Reference values captured from the Python difflib-based implementation.
func TestSemanticKnownValues(t *testing.T) {
	cases := []struct {
		output, expected string
		want             float64
	}{
		{"cancel anytime", "you can cancel anytime", 0.6389},
		{"Yes, you can cancel your subscription at any time; it ends at the cycle close.", "Yes, you can cancel your subscription at any time.", 0.6573},
		{"hello world", "world hello", 0.7273},
		{"the quick brown fox", "a quick brown dog", 0.5556},
		{"aaaa", "aa", 0.3333},
	}
	for _, c := range cases {
		approxEqual(t, SemanticSimilarity(c.output, c.expected), c.want)
	}
}

func TestSemanticStaysInUnitRange(t *testing.T) {
	got := SemanticSimilarity("cancel anytime", "you can cancel anytime")
	if got < 0.0 || got > 1.0 {
		t.Fatalf("semantic score %v out of [0,1]", got)
	}
}

func TestGetScorerReturnsEachBuiltin(t *testing.T) {
	for _, name := range []string{"exact_match", "contains", "regex", "semantic"} {
		fn, err := GetScorer(name)
		if err != nil {
			t.Fatalf("GetScorer(%q) unexpected error: %v", name, err)
		}
		if fn == nil {
			t.Fatalf("GetScorer(%q) returned nil func", name)
		}
	}
}

func TestGetScorerUnknownErrors(t *testing.T) {
	_, err := GetScorer("nonsense")
	if err == nil {
		t.Fatalf("GetScorer(nonsense) should error")
	}
}

func TestIsKnownScorer(t *testing.T) {
	if !IsKnownScorer("semantic") {
		t.Fatalf("semantic should be known")
	}
	if IsKnownScorer("bogus") {
		t.Fatalf("bogus should not be known")
	}
}

func TestScorerNamesReturnsCopy(t *testing.T) {
	names := ScorerNames()
	if len(names) != 4 {
		t.Fatalf("expected 4 scorer names, got %d", len(names))
	}
	names[0] = "mutated"
	if ScorerNames()[0] == "mutated" {
		t.Fatalf("ScorerNames must return a defensive copy")
	}
}

File: ts/src/scorers.test.ts
import { describe, expect, it } from "vitest";
import {
  contains,
  exactMatch,
  getScorer,
  isKnownScorer,
  regexMatch,
  round4,
  SCORER_NAMES,
  scorerNames,
  semanticSimilarity,
} from "./scorers";

describe("exactMatch", () => {
  it("scores identical strings 1.0", () => {
    expect(exactMatch("hello", "hello")).toBe(1.0);
  });
  it("trims surrounding whitespace on both sides", () => {
    expect(exactMatch("  hello  ", "hello")).toBe(1.0);
    expect(exactMatch("hello", "\thello\n")).toBe(1.0);
  });
  it("scores different strings 0.0", () => {
    expect(exactMatch("hello", "world")).toBe(0.0);
  });
  it("is case sensitive", () => {
    expect(exactMatch("Hello", "hello")).toBe(0.0);
  });
});

describe("contains", () => {
  it("finds a substring", () => {
    expect(contains("the quick brown fox", "quick")).toBe(1.0);
  });
  it("is case insensitive", () => {
    expect(contains("The Quick Brown FOX", "quick brown")).toBe(1.0);
  });
  it("trims the expected needle", () => {
    expect(contains("refund in 30 days", "  30 days  ")).toBe(1.0);
  });
  it("scores 0.0 when absent", () => {
    expect(contains("the quick brown fox", "lazy dog")).toBe(0.0);
  });
  it("matches an empty needle", () => {
    expect(contains("anything", "")).toBe(1.0);
  });
});

describe("regexMatch", () => {
  it("matches anywhere in the output", () => {
    expect(regexMatch("abc123def", "\\d+")).toBe(1.0);
  });
  it("honours a leading (?i) inline flag", () => {
    expect(regexMatch("NO WAY", "(?i)no")).toBe(1.0);
    expect(regexMatch("Definitely not", "(?i)(no|not|don't|do not)")).toBe(1.0);
  });
  it("scores 0.0 on no match", () => {
    expect(regexMatch("abcdef", "\\d+")).toBe(0.0);
  });
  it("respects anchors", () => {
    expect(regexMatch("xabc", "^abc")).toBe(0.0);
    expect(regexMatch("abc", "^abc$")).toBe(1.0);
  });
});

describe("semanticSimilarity", () => {
  it("scores identical strings 1.0", () => {
    expect(semanticSimilarity("abc def", "abc def")).toBe(1.0);
  });
  it("scores empty output 0.0", () => {
    expect(semanticSimilarity("", "x")).toBe(0.0);
  });
  it("scores empty expected 0.0", () => {
    expect(semanticSimilarity("x", "")).toBe(0.0);
  });
  it("scores both empty 0.0", () => {
    expect(semanticSimilarity("", "")).toBe(0.0);
  });
  it("reproduces the reordered-token reference value", () => {
    expect(semanticSimilarity("hello world", "world hello")).toBeCloseTo(0.7273, 10);
  });
  it("reproduces the partial-overlap reference value", () => {
    expect(semanticSimilarity("the quick brown fox", "a quick brown dog")).toBeCloseTo(0.5556, 10);
  });
  it("reproduces the repeated-char reference value", () => {
    expect(semanticSimilarity("aaaa", "aa")).toBeCloseTo(0.3333, 10);
  });
  it("reproduces the cancellation reference value", () => {
    const out = "Yes, you can cancel your subscription at any time; it ends at the cycle close.";
    const exp = "Yes, you can cancel your subscription at any time.";
    expect(semanticSimilarity(out, exp)).toBeCloseTo(0.6573, 10);
  });
  it("reproduces the short reference value", () => {
    expect(semanticSimilarity("cancel anytime", "you can cancel anytime")).toBeCloseTo(0.6389, 10);
  });
  it("stays within [0,1]", () => {
    const v = semanticSimilarity("the quick brown fox jumps", "the lazy dog sleeps soundly");
    expect(v).toBeGreaterThanOrEqual(0.0);
    expect(v).toBeLessThanOrEqual(1.0);
  });
});

describe("round4", () => {
  it("rounds to four decimal places", () => {
    expect(round4(1 / 3)).toBe(0.3333);
    expect(round4(0.123449)).toBe(0.1234);
  });
});

describe("scorer registry", () => {
  it("resolves every built-in scorer", () => {
    for (const name of SCORER_NAMES) {
      expect(typeof getScorer(name)).toBe("function");
    }
  });
  it("throws for an unknown scorer", () => {
    expect(() => getScorer("nope")).toThrow(/unknown scorer 'nope'/);
  });
  it("reports membership", () => {
    expect(isKnownScorer("semantic")).toBe(true);
    expect(isKnownScorer("exact_match")).toBe(true);
    expect(isKnownScorer("fuzzy")).toBe(false);
  });
  it("lists all four names in registration order", () => {
    expect(scorerNames()).toEqual(["exact_match", "contains", "regex", "semantic"]);
  });
});
